crates/tree-morph/src/events.rs

```diff
@@ -0,0 +1,35 @@
+use uniplate::Uniplate;
+
+pub(crate) struct EventHandlers<T: Uniplate, M> {
+    on_enter: Vec<fn(&T, &mut M)>,
+    on_exit: Vec<fn(&T, &mut M)>,
+}
+
+impl<T: Uniplate, M> EventHandlers<T, M> {
+    pub(crate) fn new() -> Self {
+        EventHandlers {
+            on_enter: vec![],
+            on_exit: vec![],
+        }
+    }
+
+    pub(crate) fn trigger_on_enter(&self, node: &T, meta: &mut M) {
+        for f in self.on_enter.iter() {
+            f(node, meta)
+        }
+    }
+
+    pub(crate) fn trigger_on_exit(&self, node: &T, meta: &mut M) {
```

Comment:
I think it is important to be specific in the method name on whether this applies post or pre exit.